### PR TITLE
Remove SocketAsyncEventArgs._ptrNativeOverlapped

### PR DESCRIPTION
**Pull request by stephentoub**
`SocketAsyncEventArgs` uses a `SafeNativeOverlapped` to represent the `NativeOverlapped*` it gets back from the `PreAllocatedOverlapped`.  This leads to several costs:
- Allocation/disposal of the `SafeNativeOverlapped`.  It used to be allocated/disposed for every operation, and that was fixed.  But it's still allocated/disposed every time the `SocketAsyncEventArgs` changes associated with a `Socket`, and that's problematic in a pooling situation, where a `SocketAsyncEventArgs` is taken from a pool, used for an operation, and then returned to the pool to be used subsequently by another operation, likely on a different socket.
- It adds a field to `SocketAsyncEventArgs`, as the `SafeNativeOverlapped` instance needs to be disposed of only when the operation completes.
- As with any `SafeHandle`, it adds AddRef/Release costs to each P/Invoke with which it's used.

But, it's unnecessary on `SocketAsyncEventArgs`.  The overlapped has a very controlled lifetime: it's created internally, it's passed to an internal async operation, and upon completion of that operation it's freed.  The benefits of the `SafeHandle` don't apply here:
- `SafeHandle`s help guaranteed cleanup via a finalizer, but we're not publishing this pointer and have knowledge of its complete lifetime.  The only place it would help is if we got an asynchronous exception (e.g. a thread abort) in a very unexpected place, and we don't need to be reliable against such things (nor would we be even if we added that).
- `SafeHandle`s help guarantee that the pointer won't be disposed of while still in use, but that's mainly relevant when handed out to multiple pieces of code, and in particular code we don't control... in this case, it's not handed out to any code other than that in this one file.

On my machine, this improves the throughput by ~13% in the example at https://github.com/dotnet/corefx/pull/17237#issuecomment-287613725 (which is stressing this case by changing the socket on every send/receive and doing very fast send/receives), and makes it allocation-free to change the Socket with which a SocketAsyncEventArgs is associated (whereas today there’s a 40byte SafeHandle allocation each time it changes).  With a small modification to that test to not change the buffer and instead set it once up front, this still has an ~8% throughput improvement, most of which is due to the reduction in SafeHandle marshaling and associated AddRef/Release costs.

cc: @geoffkizer, @davidsh, @cipop, @jkotas, @vancem

### CODE REVIEW

**Comment by geoffkizer** [comment]
Dumb question: is SafeNativeOverlapped.cs not getting deleted?  Or did I miss it?

**Comment by stephentoub** [comment]
> Dumb question: is SafeNativeOverlapped.cs not getting deleted? Or did I miss it?

It's still used by the APM implementations in System.Net.Sockets; I've not gone through to look at what it would mean to remove it from there.  The file is also used by the System.Net.NetworkInformation lib.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please

File: src/Common/src/Interop/Windows/Winsock/Interop.WSAIoctl.cs [hunk 1]
```diff
@@ -32,7 +32,7 @@ internal static partial class Winsock
             [Out] byte[] outBuffer,
             [In] int outBufferSize,
             [Out] out int bytesTransferred,
-            [In] SafeHandle overlapped,
+            [In] IntPtr overlapped,
             [In] IntPtr completionRoutine);
 
         [DllImport(Interop.Libraries.Ws2_32, SetLastError = true, EntryPoint = "WSAIoctl")]
```
Comment (geoffkizer) [inline src/Common/src/Interop/Windows/Winsock/Interop.WSAIoctl.cs]: SafeHandle here should probably be changed as well -- though on quick inspection, I think this method is unused and should just be removed.

**Comment by geoffkizer** [comment]
> It's still used by the APM implementations in System.Net.Sockets; 

Ah, sorry, makes sense.

> The file is also used by the System.Net.NetworkInformation lib.

Yeah, looks like you removed all the references from System.Net.NetworkInformation, though, right?

File: src/Common/src/Interop/Windows/Winsock/Interop.WSAIoctl.cs (hunk 1 above)
Comment (geoffkizer) [inline src/Common/src/Interop/Windows/Winsock/Interop.WSAIoctl.cs]: To be clear, I mean on WSAIoctl_Blocking_Internal.

Comment (stephentoub) [inline src/Common/src/Interop/Windows/Winsock/Interop.WSAIoctl.cs]: I searched and it's not used anywhere.  I'll delete it.

**Comment by stephentoub** [comment]
> Yeah, looks like you removed all the references from System.Net.NetworkInformation, though, right?

You're right, will remove it from the System.Net.NetworkInformation.csproj.

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs [hunk 2]
```diff
@@ -95,14 +95,17 @@ private static unsafe void CompletionPortCallback(uint errorCode, uint numBytes,
                     }
                     else
                     {
+                        bool addRefedNativeOverlapped = false;
                         try
                         {
+                            asyncResult._nativeOverlapped.DangerousAddRef(ref addRefedNativeOverlapped);
```
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Why is this needed now when it wasn't before?

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs [hunk 3]
```diff
@@ -139,18 +149,13 @@ private void CompletionCallback(int numBytes, SocketError socketError)
...
-            {
-                // On WinNT we need to use (due to the current implementation)
-                // an Overlapped object in order to bind the socket to the
-                // ThreadPool's completion port, so return the native handle
-                return _nativeOverlapped == null ? SafeNativeOverlapped.Zero : _nativeOverlapped;
-            }
-        }
+        // The following property returns a SafeHandle for the pointer to
+        // the Overlapped structure we're using for IO.
+        internal SafeNativeOverlapped OverlappedHandle =>
```
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: I think you should be able to remove this -- all uses of this should be replaced with the new DangerousOverlappedPointer below.

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Before it was there, it was just implicit as part of marshaling the _nativeOverlapped as a SafeHandle.  Now that the P/Invoke is taking the raw pointer instead of a SafeHandle, the runtime isn't doing this implicitly, so I added doing it explicitly.

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs [hunk 4]
```diff
@@ -139,18 +149,13 @@ private void CompletionCallback(int numBytes, SocketError socketError)
...
+            _nativeOverlapped ?? SafeNativeOverlapped.Zero;
+
+        internal unsafe NativeOverlapped* DangerousOverlappedPointer =>
+            _nativeOverlapped != null ? (NativeOverlapped*)_nativeOverlapped.DangerousGetHandle() : null;
```
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: I realize this is how the code worked previously, but.... is there ever a case where _nativeOverlapped can be null here?

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 2 above)
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Yeah, I guess my question is really, is this necessary?  What does it protect against?

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Yup.

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 2 above)
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: And unless I'm missing something, you're not doing this when we pass the overlapped in general, e.g. to WSASend etc.  So why just here?

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: It protects against the SafeNativeOverlapped being disposed of while the P/Invoke is in progress; if that were to happen, the disposal would be deferred until after the call.

I've not gone through the APM stuff to analyze whether that's actually a possibility.  That could be done as part of determining if/how to rip SafeNativeOverlapped out entirely, but I don't want to do that now.  This is a measurable cost in SAEA, but there are so many other overheads in the APM implementation, it's not as important there, and I'd like to focus our efforts on the thing we can make as fast as possible.

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Hmm.  Doesn't look like it.  Will simplify.

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: > And unless I'm missing something, you're not doing this when we pass the overlapped in general, e.g. to WSASend etc. So why just here?

You mean the ones I commented with  "// SafeHandle was just created in SetUnmanagedStructures"?  The comment was meant to highlight that the reason the "Dangerous" operation is actually safe is because the SafeHandle was just created and hasn't been given out to anyone, therefore no one could be disposing of it concurrently.

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Ok, I understand now, I think.  You're saying that using DangerousOverlappedPointer in SocketPal is okay because we can be sure Dispose hasn't been called there, but we can't here, because we are not sure there's not a path where we could have disposed by here.  Is that correct?  Seems reasonable.

That said, 
(1) We should add a comment explaining that we should probably get rid of SafeNativeOverlapped usage in BaseOverlappedAsyncResult eventually, but we need to analyze all the possible usage scenarios and make sure there's no weird paths
(2) In this particular case, I think we should just use the passed in nativeOverlapped instead of the one on the asyncResult.

**Comment by geoffkizer** [comment]
There should be some dead code on SafeNativeOverlapped now (ReplaceHandle, FreeNativeOverlapped) -- can we remove all that?

**Comment by stephentoub** [comment]
> can we remove all that?

Already did, just waiting for the existing outerloop leg to finish before I push my "remove dead code" commit :)

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: > Is that correct? Seems reasonable.

Correct

> We should add a comment explaining that we should probably get rid of SafeNativeOverlapped usage

Will add

> I think we should just use the passed in nativeOverlapped instead of the one on the asyncResult

Yup, will change.  I hadn't noticed that was being passed in to the method.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs [hunk 5]
```diff
@@ -333,40 +353,48 @@ private void InnerStartOperationReceiveFrom()
...
-                    handle,
-                    _wsaBufferArray,
-                    _bufferListInternal.Count,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                FreeNativeOverlappedIfNotPending(overlapped, socketError);
```
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: The pattern we use in SocketPal.Windows.cs to handle the equivalent situation is that we catch, free the overlapped, and rethrow.  See code below.

I don't have a strong opinion about which is better, and we don't need to resolve that as part of this checkin, but it would be nice to pick one and be consistent.

```
            asyncResult.SetUnmanagedStructures(buffer, offset, count, null);
            try
            {
                // This can throw ObjectDisposedException.
                int bytesTransferred;
                SocketError errorCode = Interop.Winsock.WSASend(
                    handle,
                    ref asyncResult._singleBuffer,
                    1, // There is only ever 1 buffer being sent.
                    out bytesTransferred,
                    socketFlags,
                    asyncResult.OverlappedHandle,
                    IntPtr.Zero);

                return asyncResult.ProcessOverlappedResult(errorCode == SocketError.Success, bytesTransferred);
            }
            catch
            {
                asyncResult.ReleaseUnmanagedStructures();
                throw;
            }
```

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: > The pattern we use in SocketPal.Windows.cs to handle the equivalent situation is that we catch, free the overlapped, and rethrow. See code below

The cases are similar but slightly different, e.g. in your example the overlapped is stored in the IAsyncResult instance, but in the SAEA case, it's only on the stack, so it would need to be passed around.  I actually initially did that, but I found it cleaner to switch to this finally model.  We can certainly revisit one or the other subsequently.

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: Not sure I understand what you mean by "it would need to be passed around".

Couldn't you just change the finally to this:

catch 
{
    FreeNativeOverlapped(nativeOverlapped);
    throw;
}

?  Again, not sure which is better necessarily.

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: Ah wait, are you saying this covers the sync completion case too?

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: Who frees it then on the success/synchronous completion path?

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: > Ah wait, are you saying this covers the sync completion case too?

Yes

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: Yeah, I get it now.

To match the SocketPal pattern, you could pass the NativeOverlapped* to ProcessIOCPResult and free it on sync completion.  We're already checking for sync completion there, so we know whether to release the overlapped or not.  With the current code you're checking for IOPending again, which is almost certainly immaterial, but strictly speaking unnecessary.

Again, not sure it matters...

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: > To match the SocketPal pattern, you could pass the NativeOverlapped* to ProcessIOCPResult and free it on sync completion

That's what I had initially (and is what I meant by passing the overlapped around), but in ProcessIOCPResult I then had to insert the Free call at a couple of different places, and I was also concerned there may be some other points where an exception could emerge and cause it not to be freed.  Previously that would have at least been handled by the SocketAsyncEventArgs.Dispose, because it would have freed the stored overlapped then, but with this change, it would simply be leaked.  So I opted for the finally approach.

If it ends up mattering, we can revisit.

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: Yeah, makes sense.

There's a lot more that happens in ProcessIOCPResult for SAEA than happens in ProcessOverlappedResult for IAsyncResult.  The intent is that the latter never throws, though looking at it now, it's not clear to me that there aren't some weird edge cases where it could throw...

Regardless:
(1) This code looks fine
(2) Longer term I think we should strive to have the minimal amount of code inside the try/catch -- i.e. just invoke the native call and process the result error code.  
(3) The only reason we have try/catch at all here is if the SafeCloseSocket marshalling throws -- if we ever revisit how this is handled we may be able to get rid of the try/catch entirely

**Comment by geoffkizer** [comment]
It's great to see this going in.  LGTM!

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please
@dotnet-bot test Outerloop Win10 Debug please
@dotnet-bot test Outerloop Ubuntu14.04 Debug please

**Comment by stephentoub** [comment]
@dotnet-bot help

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows 10 Debug please

**Comment by stephentoub** [comment]
Thanks for the review, @geoffkizer.
